contracts/tips/src/contract.rs

```diff
@@ -0,0 +1,2202 @@
+            id: msg.subspace_id.u64(),
+            error,
+        })?;
+
+    let service_fee = if let Some(service_fee) = msg.service_fee {
+        Some(StateServiceFee::try_from(service_fee)?)
+    } else {
+        None
+    };
+
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+    CONFIG.save(
+        deps.storage,
+        &Config {
+            admin,
+            subspace_id: msg.subspace_id.u64(),
+            service_fee,
+            tips_history_size: msg.tips_history_size,
+        },
+    )?;
+    // Initialize the block index.
+    BLOCK_INDEX.save(deps.storage, &(0, 0))?;
+
+    Ok(Response::new()
+        .add_attribute(ATTRIBUTE_ACTION, ACTION_INSTANTIATE)
+        .add_attribute(ATTRIBUTE_SENDER, info.sender)
+        .add_attribute(ATTRIBUTE_ADMIN, msg.admin)
+        .add_attribute(ATTRIBUTE_SUBSPACE_ID, msg.subspace_id)
+        .add_attribute(
+            ATTRIBUTE_TIPS_HISTORY_SIZE,
+            msg.tips_history_size.to_string(),
+        ))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut<DesmosQuery>,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response<DesmosMsg>, ContractError> {
+    msg.validate()?;
+
+    match msg {
+        ExecuteMsg::SendTip { target, amount } => execute_send_tip(deps, env, info, target, amount),
+        ExecuteMsg::UpdateServiceFee { new_fee } => execute_update_service_fee(deps, info, new_fee),
+        ExecuteMsg::UpdateAdmin { new_admin } => execute_update_admin(deps, info, new_admin),
+        ExecuteMsg::UpdateSavedTipsHistorySize { new_size } => {
+            execute_update_saved_tips_history_size(deps, info, new_size)
+        }
+        ExecuteMsg::ClaimFees { receiver } => execute_claim_fees(deps, env, info, receiver),
+    }
+}
+
+fn execute_send_tip(
+    deps: DepsMut<DesmosQuery>,
+    env: Env,
+    info: MessageInfo,
+    target: Target,
+    tip_amount: Vec<Coin>,
+) -> Result<Response<DesmosMsg>, ContractError> {
+    if info.funds.is_empty() {
+        return Err(ContractError::EmptyFunds {});
+    }
+
+    let config = CONFIG.load(deps.storage)?;
+
+    if let Some(service_fee) = &config.service_fee {
+        service_fee.check_fees(&info.funds, &tip_amount)?;
+    }
+
+    // Resolve the receiver and the optional post id
+    let (post_id, receiver) = match target {
+        Target::ContentTarget { post_id } => {
+            let post = PostsQuerier::new(deps.querier.deref())
+                .query_post(config.subspace_id, post_id.u64())
+                .map_err(|_| ContractError::PostNotFound { id: post_id.u64() })?
+                .post;
+            (post.id.u64(), post.author)
+        }
+        Target::UserTarget { receiver } => {
+            // Set the post id to 0 (invalid id) to signal that this tip is referencing an user
+            (0_u64, deps.api.addr_validate(&receiver)?)
+        }
+    };
+
+    if info.sender == receiver {
```

Comment:
Can we also check that the sender has a profile? I would avoid profileless people to tip inside our app. Then if someone wants to have the contract without this check it can simply remove it and compile a new one